plugins/uefi-capsule/fu-uefi-grub-device.c

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (C) 2018 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-uefi-common.h"
+#include "fu-uefi-grub-device.h"
+
+struct _FuUefiGrubDevice {
+	FuUefiDevice parent_instance;
+};
+
+G_DEFINE_TYPE(FuUefiGrubDevice, fu_uefi_grub_device, FU_TYPE_UEFI_DEVICE)
+
+static gboolean
+fu_uefi_grub_device_mkconfig(FuDevice *device,
+			     const gchar *esp_path,
+			     const gchar *target_app,
+			     GError **error)
+{
+	const gchar *argv_mkconfig[] = {"", "-o", "/boot/grub/grub.cfg", NULL};
+	const gchar *argv_reboot[] = {"", "fwupd", NULL};
+	g_autofree gchar *grub_mkconfig = NULL;
+	g_autofree gchar *grub_reboot = NULL;
+	g_autofree gchar *grub_target = NULL;
+	g_autofree gchar *localstatedir = fu_common_get_path(FU_PATH_KIND_LOCALSTATEDIR_PKG);
+	g_autofree gchar *output = NULL;
+	g_autoptr(GString) str = g_string_new(NULL);
+
+	/* find grub.conf */
+	if (!g_file_test(argv_mkconfig[2], G_FILE_TEST_EXISTS))
+		argv_mkconfig[2] = "/boot/grub2/grub.cfg";
+	if (!g_file_test(argv_mkconfig[2], G_FILE_TEST_EXISTS)) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "could not find grub.conf");
+		return FALSE;
+	}
+
+	/* find grub-mkconfig */
+	grub_mkconfig = fu_common_find_program_in_path("grub-mkconfig", NULL);
+	if (grub_mkconfig == NULL)
+		grub_mkconfig = fu_common_find_program_in_path("grub2-mkconfig", NULL);
+	if (grub_mkconfig == NULL) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "could not find grub-mkconfig");
+		return FALSE;
+	}
+
+	/* find grub-reboot */
+	grub_reboot = fu_common_find_program_in_path("grub-reboot", NULL);
+	if (grub_reboot == NULL)
+		grub_reboot = fu_common_find_program_in_path("grub2-reboot", NULL);
+	if (grub_reboot == NULL) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "could not find grub-reboot");
+		return FALSE;
+	}
+
+	/* replace ESP info in conf with what we detected */
+	g_string_append_printf(str, "EFI_PATH=%s\n", target_app);
+	fu_common_string_replace(str, esp_path, "");
+	g_string_append_printf(str, "ESP=%s\n", esp_path);
+	grub_target = g_build_filename(localstatedir, "uefi_capsule.conf", NULL);
+	if (!g_file_set_contents(grub_target, str->str, -1, error))
+		return FALSE;
+
+	/* refresh GRUB configuration */
+	argv_mkconfig[0] = grub_mkconfig;
+	if (!g_spawn_sync(NULL,
+			  (gchar **)argv_mkconfig,
+			  NULL,
+			  G_SPAWN_DEFAULT,
+			  NULL,
+			  NULL,
+			  &output,
+			  NULL,
+			  NULL,
+			  error))
+		return FALSE;
+	g_debug("%s", output);
```

Comment:
do you think we need this debugging still?  I was pretty useful early on, but I think we only really care about logging for failures.

Reply:
I've hidden it with a FWUPD_UPDATE_CAPSULE_VERBOSE